### DISCUSSION TITLE
求助，websocket.connect 创建第28234个socket必失败，ulimit 已经改过了。

### DISCUSSION THREAD

**Discussion by lsx1994** [cloudwu/skynet#1542] [Q&A] [answered]
系统是 ubuntu server lts 20.0.4，CPU8核，内存16G，运行机器人时CPU和内存只占用一半，ulimit  n 显示 1000000。  
  
机器人使用自带的websocket.connect 发起连接，报错时在另一个机器继续开，服务器是可以继续接收连接的。    
  
skynet更新到最新主干：[eaba8f9](https://github.com/cloudwu/skynet/commit/eaba8f969c177aaf190bce40723e55827452d90c)  
  
不管重启多少次机器人，都是在28234的时候失败 ，报错log：
<pre> 
24/02/22 22:10:58.44 [:00006e53] LAUNCH snlua robot/RobotClient 28233
24/02/22 22:10:58.45 [:00006e53] init service failed: ./skyframe/engine/lualib/http/sockethelper.lua:98: socket_error5
stack traceback:
	[C]: in function 'error'
	./skyframe/engine/lualib/http/sockethelper.lua:98: in function 'http.sockethelper.connect'
	./skyframe/engine/lualib/http/websocket.lua:448: in function 'skyframe.engine.lualib.http.websocket.connect'
	./skyframe/engine/develop/net/WSClient.lua:32: in method 'Connect'
	./robot/module/RobotLogin.lua:15: in method 'RobotStart'
	./robot/RobotModule.lua:28: in method 'RobotStart'
	./robot/RobotClient.lua:33: in method 'Run'
	./skyframe/engine/develop/BaseServer.lua:58: in upvalue 'start'
	./skyframe/engine/lualib/skynet.lua:935: in function <./skyframe/engine/lualib/skynet.lua:933>
	[C]: in function 'xpcall'
	./skyframe/engine/lualib/skynet.lua:937: in function 'skynet.init_service'
	./skyframe/engine/lualib/skynet.lua:950: in upvalue 'f'
	./skyframe/engine/lualib/skynet.lua:253: in function <./skyframe/engine/lualib/skynet.lua:252>
24/02/22 22:10:58.45 [:00006e53] KILL self
24/02/22 22:10:58.45 [:00000008] init service failed: ./skyframe/engine/lualib/skynet.lua:590: call failed
stack traceback:
	[C]: in function 'error'
	./skyframe/engine/lualib/skynet.lua:590: in upvalue 'yield_call'
	./skyframe/engine/lualib/skynet.lua:607: in function 'skynet.call'
	(...tail calls...)
	./robot/RobotMain.lua:21: in upvalue 'start'
	./skyframe/engine/lualib/skynet.lua:935: in function <./skyframe/engine/lualib/skynet.lua:933>
	[C]: in function 'xpcall'
	./skyframe/engine/lualib/skynet.lua:937: in function 'skynet.init_service'
	./skyframe/engine/lualib/skynet.lua:950: in upvalue 'f'
	./skyframe/engine/lualib/skynet.lua:253: in function <./skyframe/engine/lualib/skynet.lua:252>
  </pre>

**Comment by firedtoad** [comment]
检查一下是不是内存不够了？

lsx1994 ***@***.***> 于2022年2月24日周四 17:50写道：

**Reply by lsx1994** [reply]
还有一半的内存

**Comment by aasdpiao** [comment]
一般都还是要标注一下skynet的版本号吧. 或者更新到最新版本看看. 
感觉能启动到20000+应该并非代码问题,一般是系统级别的限制. 或者是一些内存泄露引起的吧

**Reply by lsx1994** [reply]
下午刚更新了最新版， 还是一样

**Reply by sniper00** [reply]
> 下午刚更新了最新版， 还是一样

建议打印下socket error 具体内容

**Reply by lsx1994** [reply]
> > 下午刚更新了最新版， 还是一样
> 
> 建议打印下socket error 具体内容

框架统一打印[Socket Error]，我加了序号重新跑了一遍

**Comment by cloudwu** [comment]
用图片贴文本的帖子不回复。

**Reply by lsx1994** [reply]
不好意思我改好了，我以为只有代码要贴文本。

**Comment by cloudwu** [comment, marked as answer]
https://github.com/cloudwu/skynet/blob/master/lualib/http/sockethelper.lua#L93 此处 socket.open 返回 nil 连接失败时，还有第二个返回值是出错信息。建议输出出来看看失败原因。

**Reply by lsx1994** [reply]
返回值是 `Cannot assign requested address`
  
我把/etc/sysctl.conf 的 net.ipv4.ip_local_port_range 改成 10000     65000 就可以了，谢谢。
